### PR TITLE
Fix: OpenTelemetry access logs: Missing span ID

### PR DESCRIPTION
**Pull request by ashishb-solo**
Commit Message: Fix missing span_id in opentelemetry access log messages
Additional Description:
Risk Level: Low
Testing: Covered by existing tests (see ](https://github.com/envoyproxy/envoy/pull/33909/files#diff-6f359b1fb4a03ceb3f64ed506c57b32b67e06f8dc41f396f6f20c9b4f292e39f))
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] Fixes https://github.com/envoyproxy/envoy/issues/33906
Fixes https://github.com/envoyproxy/envoy/issues/30268
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc [hunk 1]
```diff
@@ -236,6 +237,8 @@ void Span::injectContext(Tracing::TraceContext& trace_context, const Tracing::Up
   }
 }
 
+std::string Span::getSpanIdAsHex() const { return std::string{}; }
```
Comment (adisuissa) [inline source/extensions/tracers/opencensus/opencensus_tracer_impl.cc:240]: Prefer returning EMPTY_STRING.

File: source/extensions/tracers/datadog/span.cc [hunk 2]
```diff
@@ -137,11 +137,16 @@ void Span::setBaggage(absl::string_view, absl::string_view) {
...
 std::string Span::getTraceIdAsHex() const {
   if (!span_) {
-    return std::string{};
+    return EMPTY_STRING;
   }
   return absl::StrCat(absl::Hex(span_->id()));
 }
 
+std::string Span::getSpanIdAsHex() const {
+  // TODO not implemented
```
Comment (adisuissa) [inline source/extensions/tracers/datadog/span.cc:146]: Please add a name/PR# and consider adding some context on what needs to be added.

File: source/extensions/access_loggers/open_telemetry/access_log_impl.cc [hunk 3]
```diff
@@ -102,6 +102,8 @@ void AccessLog::emitLog(const Formatter::HttpFormatterContext& log_context,
     auto trace_id = absl::StrCat(Hex::uint64ToHex(0), trace_id_hex);
     *log_entry.mutable_trace_id() = absl::HexStringToBytes(trace_id);
   }
+  std::string span_id_hex = log_context.activeSpan().getSpanIdAsHex();
```
Comment (adisuissa) [inline source/extensions/access_loggers/open_telemetry/access_log_impl.cc:105]: nit: const

File: changelogs/current.yaml [hunk 4]
```diff
@@ -30,6 +30,9 @@ bug_fixes:
   change: |
     Fix a RELEASE_ASSERT when using :ref:`auto_sni <envoy_v3_api_field_config.core.v3.UpstreamHttpProtocolOptions.auto_sni>`
     if the downstream request ``:authority`` was longer than 255 characters.
+- area: tracing
+  change: |
+    Fix an issue where span id is missing from opentelemetry access log entries
```
Comment (adisuissa) [inline changelogs/current.yaml:35]: style-nit: add a '.' at the end of the line (also for comments in the source files).

**Review by adisuissa — COMMENTED** [review]
Thanks for fixing this!
Assigning OTEL/tracing code owners.
/assign @yanavlasov @wbpcode

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc (hunk 1 above)
Comment (ashishb-solo) [reply source/extensions/tracers/opencensus/opencensus_tracer_impl.cc]: Thanks, fixed that [here](https://github.com/envoyproxy/envoy/pull/33909/commits/2f9767d7f6f7edbbcacd148a2efc31953b87df2b#diff-e70df531b553c85beb87a955c2f42c81df544610d09492b4037802c4a85d34b6R240).

File: source/extensions/tracers/datadog/span.cc (hunk 2 above)
Comment (ashishb-solo) [reply source/extensions/tracers/datadog/span.cc]: I addressed this comment in [this commit](https://github.com/envoyproxy/envoy/pull/33909/commits/0639045524e6f1cc95a4526c50037a45989aa41d). To be completely truthful, I have no experience with Datadog, so I don't believe I have any more additional context to add. Please let me know if there's anything else I can add here - thanks!

File: source/extensions/access_loggers/open_telemetry/access_log_impl.cc (hunk 3 above)
Comment (ashishb-solo) [reply source/extensions/access_loggers/open_telemetry/access_log_impl.cc]: I fixed this issue [here](https://github.com/envoyproxy/envoy/pull/33909/commits/2f9767d7f6f7edbbcacd148a2efc31953b87df2b#diff-558f961772efddc0955235c7588402f93ee2691724bc833ead18acfac6d0b83fL105).

File: changelogs/current.yaml (hunk 4 above)
Comment (ashishb-solo) [reply changelogs/current.yaml]: I checked through all source changes and am pretty sure that I have fixed all instances of comments that were not terminated with periods.

**Comment by ashishb-solo** [comment]
@mmanciop can you advise on whether this fixes https://github.com/envoyproxy/envoy/issues/30268? I thought it did at first glance, but now I'm not completely sure

**Comment by mmanciop** [comment]
> @mmanciop can you advise on whether this fixes https://github.com/envoyproxy/envoy/issues/30268? I thought it did at first glance, but now I'm not completely sure

I'll test it out with Dash0 :-)

**Comment by mmanciop** [comment]
It does fix it, gj :-)

**Review by mmanciop — APPROVED** [review]

**Comment by phlax** [comment]
@kyessenov i think this is waiting on further review

**Comment by ravenblackx** [comment]
@kyessenov Second ping on this waiting for your review.

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.h [hunk 5]
```diff
@@ -85,6 +85,9 @@ class ZipkinSpan : public Tracing::Span {
 
   std::string getTraceIdAsHex() const override { return span_.traceIdAsHexString(); };
 
+  // TODO: This method is unimplemented for Zipkin.
```
Comment (kyessenov) [inline source/extensions/tracers/zipkin/zipkin_tracer_impl.h:88]: nit: a link to an issue with the follow-up work would be better (here and other stubs).

File: source/extensions/access_loggers/open_telemetry/access_log_impl.cc [hunk 6]
```diff
@@ -102,6 +102,8 @@ void AccessLog::emitLog(const Formatter::HttpFormatterContext& log_context,
     auto trace_id = absl::StrCat(Hex::uint64ToHex(0), trace_id_hex);
     *log_entry.mutable_trace_id() = absl::HexStringToBytes(trace_id);
   }
+  const std::string span_id_hex = log_context.activeSpan().getSpanIdAsHex();
+  *log_entry.mutable_span_id() = absl::HexStringToBytes(span_id_hex);
```
[the remark this answers is not in the document]
Comment (ashishb-solo) [reply source/extensions/access_loggers/open_telemetry/access_log_impl.cc]: I addressed this in: https://github.com/envoyproxy/envoy/pull/33909/commits/7344695628ace7828f6ef4220f97fc1f4c4f8c50. Let me know if that looks right. Thanks!

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.h (hunk 5 above)
Comment (ashishb-solo) [reply source/extensions/tracers/zipkin/zipkin_tracer_impl.h]: I covered this in https://github.com/envoyproxy/envoy/pull/33909/commits/7344695628ace7828f6ef4220f97fc1f4c4f8c50 - let me know if that looks better

**Comment by ashishb-solo** [comment]
@kyessenov I believe all review comments have been addressed. Would you mind re-reviewing when you get a chance? Thanks!!

**Review by kyessenov — APPROVED** [review]
Thanks, LG!

**Comment by ashishb-solo** [comment]
I'm a bit confused this this hasn't merged yet. @kyessenov , do you see any issues or things that I still need to resolve?

**Review by mmanciop — APPROVED** [review]

**Comment by adisuissa** [comment]
> I'm a bit confused this this hasn't merged yet. @kyessenov , do you see any issues or things that I still need to resolve?

Apologies!
Assigning Yan as code-owner and senior-maintainer.
/assign @yanavlasov

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**